Problem: keeping replaceable objects in sync with source files

diogob opened the pull request:
## Problem
Keeping database functions, views and other replaceable objects in sync with source files while having migrations that have to be executed only once.

This is compounded by the fact that migrations may introduce dependencies for objects in a src directory and vice-versa.

## Solution

Create 2 distinct sources of database objects:

* Commands inside a *migrations* directory are always executed in order and only once.
* Commands inside any other directory (e.g. src) will be interspersed with migrations according to their dependencies and are always executed, but require the statements to have an *OR REPLACE* clause.

We use assemble_schema to figure out dependencies in a temporary database and is_replace_statement to validate the non-migration sources.

The decision to split the functionality into a function to plan and another one to execute is to make it easier to understand, test and in the future implement dry-run or other plan manipulation functionalities.

## TODO
- [x] Handle migration errors.
- [x] How to handle files removed from `src` directories? Should we keep track of these objects and remove them as in `load_from_fs` ?
- [x] The code as is here create several record in `omni_schema.migrations` with the same `name` value. Should we aggregate statements per file? - we will leave the duplicate names for now and add some hashing to identify individual statements with ease.
- [x] Create a test case for cases such as python `requirements.txt`. But it seems that this will fail since it has no `OR REPLACE` clause. Should we create an exception for things processed by `omni_schema.auxiliary_tools` ?
- [x] Docs

diogob committed:
Problem: Keeping database functions, views and other replaceable objects in sync with source files
while having migrations that have to be executed only once.

This is compounded by the fact that migrations may introduce
dependencies for objects in a src directory and vice-versa.

Solution: Create 2 distinct sources of database objects:

* Commands inside a *migrations* directory are always executed in order and only once.
* Commands inside any other directory (e.g. src) will be interspersed
  with migrations according to their dependencies and are always
  executed, but require the statements to have an *OR REPLACE* clause.

We use assemble_schema to figure out dependencies in a temporary
database and is_replace_statement to validate the non-migration sources.

yrashk commented:
> How to handle files removed from src directories? Should we keep track of these objects and remove them as in load_from_fs ?

Gut reaction – yes. But what does add complexity-wise? Should it be part of incremental recording?

diogob committed:
Problem: assemble_schema do not keep track of what is actually executed,
so anyone using the function for planning purposes would have to
recreate the logic for statement pre-processing.

This makes it harder to implement our new `make_plan` and `make` since
they would get only the raw contents of migrations (which are not valid
statements).

Soluction: change the assemble_schema return type adding a field for
what is actually executed for each returned line.

diogob committed:
Problem: make_plan was not taking pre-processing done by assemble_schema.
This would happen with auxiliary tools and files that are not SQL.
The initial implementation would just break an bail out.

Solution: once we have assemble_schema outputting the actual executed
statement we can use this already pre-processed version to make our
plan.

As a consequence it was easier to bring syntax errors from
assemble_schema output as well instead of re-validating (as the previous
implementation was doing).

diogob committed:
Problem: Policies do not have dependencies and could be recreated.
However with the current implementation they have to be placed in the
migrations directory adding confusion and toil to maintain.

Solution: Add a special case for CREATE POLICY statements where we
inject a DROP command before their creation when creating our plan.

diogob commented:
> > How to handle files removed from src directories? Should we keep track of these objects and remove them as in load_from_fs ?
> 
> Gut reaction – yes. But what does add complexity-wise? Should it be part of incremental recording?

It introduces potential issues where removing a file either breaks the migration system because of unmet dependencies or we drop cascade and potentially create more harm. 
I thought a bit more about this, and leaving the removals to the user discretion on the migrations seems the safest way to proceed with haste and it gets the job done for now.

yrashk commented on extensions/omni_schema/src/make_plan.sql:
```diff
@@ -0,0 +1,129 @@
+create or replace function make_plan(path text) 
```
The code here assumes `local_fs`, however the intention is that we pass VFS instances to make sure it works off any source. You can see it in other places like `assemble_schema`.

`omni_vfs` interface is still a bit clunky so it is understandably a bit more difficult to work with it when it comes to remote connections. This is why `assemble_schema` copies from a local vfs into remote table_fs.

I wonder how difficult will it be for us to make it take fs at this point before rolling out the next release.

Longer-term we just need a good solution to cross-connection FS. Perhaps it might be not too difficult to have a `remote_fs` that goes over fdw?

yrashk replied:
https://github.com/omnigres/omnigres/pull/755

yrashk commented on extensions/omni_schema/docs/reference.md:
```diff
@@ -160,4 +160,50 @@ The support for languages is configurable through `omni_schema.languages` table:
+
+It is also important to note that a temporary database is created in the current cluster (and using the same credentials as the caller) in order to plan the statements to be used by `make`.
+Therefore the user executing this should have all appropriate system privileges.
+Create a blank temporary database to run assemble_schema. 
```
I am not sure I understand how the last line is related to `make`.

diogob replied:
:man_facepalming: some editing left-over

yrashk reviewed (commented):
Here's the first review pass. Great work, see some commentary!

yrashk commented on extensions/omni_schema/src/make_plan.sql:
```diff
@@ -0,0 +1,129 @@
+as
+$$
+declare
+    local_conn_info text := 'host=localhost port=' || (select setting from pg_settings where name = 'port') || ' user=' || current_user;
```
In other parts (omni_test) I've used unix port instead because I suspect it is even more commonly available in local setups. Probably not a big deal, but just as a note:

https://github.com/omnigres/omnigres/blob/master/extensions/omni_test/src/instantiate.sql#L66-L68

yrashk commented on extensions/omni_schema/src/make_plan.sql:
```diff
@@ -0,0 +1,129 @@
+
+    for migration in select * from dblink(
+        plan_db_conn,
+        format('select migration_filename, executed_statement, migration_statement, execution_error from omni_schema.assemble_schema(%L, omni_vfs.local_fs(%L))', plan_db_conn, path)
```
Would this approach mean we might not be able to ensure the order of migrations?

If a migration failed and proceeded further down migrations and only then went for sources, this would mean we'll come and try to re-apply the incremental migration.

Whereas if I got it right, the algorithm is to try migrations in order and when they error our, try to run sources.

diogob replied:
It seems to me that the `assemble_schema` order will be preserved. But having it implicit is certainly not a good idea. Perhaps we could output an order from `assemble_schema` at least for now. 

The goal was to have something quick that would not require changes in `assemble_schema`. Later it might be better to just rewrite the logic inside `make_plan`.

yrashk commented on extensions/omni_schema/src/make_plan.sql:
```diff
@@ -0,0 +1,129 @@
+        else
+            -- Non-migrations are picky, we error unless conditions are met
+            if not (
+                migration.executed_statement ~* 'select omni_python.install_requirements'
```
is this particular line fragile? or good enough for now?

diogob replied:
Yes, I was not happy with this. Was mostly a stop-gap solution. But we should review the approach to `omni_python` and other language loaders. Perhaps they should live in their own modules and we could invert the dependency, so they would use `omni_schema` to build a more refined plan.

yrashk commented:
A note on replaceability:

```postgresql
omni_httpd=# create or replace procedure print() language plpgsql as $$ begin raise notice '%', current_setting('[test.me](https://test.me/)'); end; $$;
CREATE PROCEDURE
omni_httpd=# alter procedure print() set [test.me](https://test.me/) to 'hello';
ALTER PROCEDURE
omni_httpd=# call print();
NOTICE: hello
CALL
omni_httpd=# create or replace procedure print() language plpgsql as $$ begin raise notice '%', current_setting('[test.me](https://test.me/)'); end; $$;
CREATE PROCEDURE
omni_httpd=# call print();
NOTICE: 
CALL
```

This shows that things like altered settings vanish upon replacement. Which means that straight-out replacing objects is actually not lossless.

Reading documentation:

<blockquote>
When CREATE OR REPLACE PROCEDURE is used to replace an existing procedure, the ownership and permissions of the procedure do not change. All other procedure properties are assigned the values specified or implied in the command.
</blockquote>

The language of it is kind of vague but it may suggest this is intentional, or at least known.
